libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,72 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+const DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = url.unwrap_or(DEFAULT_GOALSTATE_URL);
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            let mut rest_timeout = total_timeout;
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(http::WIRESERVER_HTTP_TIMEOUT_SEC))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-        let goalstate: Goalstate = from_str(&body)?;
-        Ok(goalstate)
-    } else {
-        Err(Error::HttpStatus {
-            endpoint: url.to_owned(),
-            status: response.status(),
-        })
-    }
+                if statuscode.is_success() && statuscode == StatusCode::OK {
+                    tracing::info!("{}", format!("HTTP response succeeded with status {}", statuscode));
+                    return Ok(response);
+                }
+
+                if !http::RETRY_CODES.contains(&statuscode) {
+                    return response.error_for_status().map_err(|err| {
+                        tracing::error!(error = ?err, "{}", format!("HTTP response failed immediately due to status {}", statuscode));
```

Comment:
In the case of timeout (connection/read), it is not "immediate"

```suggestion
                        tracing::error!(error = ?err, "{}", format!("HTTP response failed due to status {}", statuscode));
```

Reply:
Done